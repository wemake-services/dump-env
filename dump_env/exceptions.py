class StrictEnvError(Exception):
    """We use this exception when some env vars are missing from environ."""
