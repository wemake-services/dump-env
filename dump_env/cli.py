import argparse
import sys
from typing import Final, NoReturn

from dump_env import dumper
from dump_env.exceptions import StrictEnvError

# Characters that require the value to be quoted
QUOTE_CHARS: Final = (' ', '\n', '=', '"', "'")


def needs_quotes(raw_env_value: str) -> bool:
    """
    Check if the value needs to be quoted.

    Args:
        raw_env_value (str): The value to check.

    Returns:
        bool: True if the value needs to be quoted, False otherwise.
    """
    if not raw_env_value:
        return False
    return any(char in raw_env_value for char in QUOTE_CHARS)


def escape(raw_env_value: str) -> str:
    """
    Escape the value for use in an environment variable.

    Args:
        raw_env_value (str): The value to escape.

    Returns:
        str: The escaped value.
    """
    return (
        raw_env_value
        # Backslashes need to be escaped
        .replace('\\', '\\\\')  # noqa: WPS348, WPS342
        # Quotes in the value need to be escaped
        .replace('"', '\\"')  # noqa: WPS348, WPS342
    )


def _create_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        '-p',
        '--prefix',
        type=str,
        action='append',
        help='Adds prefix',
    )
    parser.add_argument(
        '-t',
        '--template',
        default='',
        type=str,
        help='Adds template path',
    )
    parser.add_argument(
        '--strict',
        type=str,
        action='append',
        help='Strict variables should exists in os envs',
    )
    parser.add_argument(
        '-s',
        '--source',
        default='',
        type=str,
        help='Source template path, restricts non-prefixed env vars',
    )
    parser.add_argument(
        '--strict-source',
        action='store_true',
        help='All source template variables should exist in os envs',
    )
    return parser


def main() -> NoReturn:
    """
    Runs dump-env script.

    Example::

        This example will dump all environ variables:

        .. code:: bash

            $ dump-env

        This example will dump all environ variables starting with ``PIP_``:

        .. code:: bash

            $ dump-env -p 'PIP_'

        This example will dump all environ variables starting with ``PIP_``
        and update them with variables starting with ``SECRET_``:

        .. code:: bash

            $ dump-env -p 'PIP_' -p 'SECRET_'

        This example will dump everything from ``.env.template`` file
        and all env variables with ``SECRET_`` prefix into a ``.env`` file:

        .. code:: bash

            $ dump-env -p 'SECRET_' -t .env.template > .env

        This example will fail if ``REQUIRED`` does not exist in environ:

        .. code:: bash

            $ dump-env --strict=REQUIRED

        This example will dump everything from a source ``.env.template`` file
        with only env variables that are defined in the file:

        .. code:: bash

            $ dump-env -s .env.template

        This example will fail if any keys in the source template do not exist
        in the environment:

        .. code:: bash

            $ dump-env -s .env.template --strict-source

    """
    args = _create_parser().parse_args()
    strict_vars = set(args.strict) if args.strict else None

    try:
        variables = dumper.dump(
            args.template,
            args.prefix,
            strict_vars,
            args.source,
            args.strict_source,
        )
    except StrictEnvError as exc:
        sys.stderr.write(f'{exc!s}\n')
        sys.exit(1)
    else:
        for env_name, env_value in variables.items():
            if needs_quotes(env_value):
                env_value = f'"{escape(env_value)}"'  # noqa: PLW2901
            sys.stdout.write(f'{env_name}={env_value}\n')
        sys.exit(0)
