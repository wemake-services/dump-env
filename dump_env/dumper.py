from collections import OrderedDict
from collections.abc import Mapping
from os import environ
from pathlib import Path
from typing import Final

from dump_env.exceptions import StrictEnvError

Store = Mapping[str, str]

EMPTY_STRING: Final = ''


def _parse(source: str) -> Store:
    """
    Reads the source ``.env`` file and load key-values.

    Args:
        source: ``.env`` template filepath

    Returns:
        Store with all keys and values.

    """
    parsed_data = {}

    with Path(source).open(encoding='utf-8') as env_file:
        for line in env_file:
            line = line.strip()  # noqa: PLW2901

            if not line or line.startswith('#') or '=' not in line:
                # Ignore comments and lines without assignment.
                continue

            # Remove whitespaces and quotes:
            env_name, env_value = line.split('=', 1)
            env_name = env_name.strip()
            env_value = env_value.strip().strip('\'"')
            parsed_data[env_name] = env_value

    return parsed_data


def _preload_existing_vars(prefix: str) -> Store:
    """Preloads env vars from environ with the given prefix."""
    if not prefix:
        # If prefix is empty just return all the env variables.
        return environ

    prefixed = {}

    # Prefix is not empty, do the search and replacement:
    for env_name, env_value in environ.items():
        if not env_name.startswith(prefix):
            # Skip vars with no prefix.
            continue

        prefixed[env_name.replace(prefix, '', 1)] = env_value

    return prefixed


def _preload_specific_vars(env_keys: set[str]) -> Store:
    """Preloads env vars from environ in the given set."""
    specified = {}

    for env_name, env_value in environ.items():
        if env_name not in env_keys:
            # Skip vars that have not been requested.
            continue

        specified[env_name] = env_value

    return specified


def _assert_envs_exist(strict_keys: set[str]) -> None:
    """Checks that all variables from strict keys do exists."""
    missing_keys: list[str] = [
        strict_key for strict_key in strict_keys if strict_key not in environ
    ]

    if missing_keys:
        raise StrictEnvError(
            'Missing env vars: {}'.format(', '.join(missing_keys)),
        )


def _source(source: str, strict_source: bool) -> Store:  # noqa: FBT001
    """Applies vars and assertions from source template ``.env`` file."""
    sourced: dict[str, str] = {}
    sourced.update(_parse(source))

    if strict_source:
        _assert_envs_exist(set(sourced.keys()))

    sourced.update(_preload_specific_vars(set(sourced.keys())))
    return sourced


def dump(
    template: str = EMPTY_STRING,
    prefixes: list[str] | None = None,
    strict_keys: set[str] | None = None,
    source: str = EMPTY_STRING,
    strict_source: bool = False,  # noqa: FBT001, FBT002
) -> dict[str, str]:
    """
    This function is used to dump ``.env`` files.

    As a source you can use both:
    1. env.template file (``''`` by default)
    2. env vars prefixed with some prefix (``''`` by default)

    Args:
        template: The path of the ``.env`` template file,
           use an empty string when there is no template file.

        prefixes: List of string prefixes to use only certain env
           variables, could be an empty string to use all available variables.

        strict_keys: List of keys that must be presented in env vars.

        source: The path of the ``.env`` template file,
           defines the base list of env vars that should be checked,
           disables the fetching of non-prefixed env vars,
           use an empty string when there is no source file.

        strict_source: Whether all keys in source template must also be
           presented in env vars.

    Returns:
        Ordered key-value pairs of dumped env and template variables.

    Raises:
        StrictEnvError: when some variable from template is missing.

    """
    if prefixes is None:
        prefixes = [] if source else [EMPTY_STRING]

    if strict_keys:
        _assert_envs_exist(strict_keys)

    store: dict[str, str] = {}

    if source:
        # Loading env values from source template file:
        store.update(_source(source, strict_source))

    if template:
        # Loading env values from template file:
        store.update(_parse(template))

    # Loading env variables from `os.environ`:
    for prefix in prefixes:
        store.update(_preload_existing_vars(prefix))

    # Sort keys and keep them ordered:
    return OrderedDict(sorted(store.items()))
