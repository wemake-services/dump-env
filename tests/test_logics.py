import pytest

from dump_env import dumper


def simple_environ(prefix='', env_value='value'):
    """Returns dict with example environment with given prefix and value."""
    return {
        f'{prefix}key': env_value,
        'a': 'b',
    }


def same_environ():
    """Returns dict with example environment with given prefix and value."""
    return {
        'NORMAL_KEY': 'test',
    }


def multiple_variables_with_prefix():
    """Returns dict with environment with multiple prefixed variables."""
    return {
        'SECRET_DJANGO_SECRET_KEY': 'test',
        'SECRET_SECRET_VALUE': 'value',
    }


@pytest.mark.usefixtures('env_file')
class TestParse:
    """Test parse function."""

    def test_parse_normal(self, env_file):
        """Ensures that given env keys are present in output."""
        parsed_data = dumper._parse(env_file)  # noqa: SLF001

        assert isinstance(parsed_data, dict)
        assert 'NORMAL_KEY' in parsed_data
        assert parsed_data['NORMAL_KEY'] == 'SOMEVALUE'

    def test_parse_exceptions(self, env_file):
        """Ensures that unknown env keys are not present in output."""
        parsed_data = dumper._parse(env_file)  # noqa: SLF001

        assert isinstance(parsed_data, dict)
        assert 'COMMENTED_KEY' not in parsed_data
        assert 'KEY_WITH_NO_ASSIGNMENT' not in parsed_data


@pytest.mark.usefixtures('env_file')
class TestFill:
    """Test fill function."""

    def test_fill_template_var(self, env_file_with_template):
        """Ensures that given env keys are fill template env var."""
        parsed_data = dumper._parse(env_file_with_template)  # noqa: WPS437
        filled_data = dumper._fill(parsed_data)  # noqa: WPS437
        assert 'VAR_TEMPLATE' in filled_data
        assert filled_data['VAR_TEMPLATE'] == 'value1 value2'

    def test_parse_variable_names(self):
        """Test template variable parser."""
        template = '${FIRST_VALUE} ${SECOND_VALUE}'
        variable_names = dumper._parse_variable_names(template)  # noqa: WPS437
        assert variable_names == ['FIRST_VALUE', 'SECOND_VALUE']

    def test_fill_wrong_template(self, env_file_with_wrong_template):
        """Ensures that an incorrectly filled template is not filled in."""
        parsed_data = dumper._parse(  # noqa: WPS437
            env_file_with_wrong_template,
        )
        filled_data = dumper._fill(parsed_data)  # noqa: WPS437
        assert parsed_data == filled_data


@pytest.mark.usefixtures('monkeypatch', 'env_file')
class TestDump:
    """Test dump function."""

    def test_with_default_arguments(self, monkeypatch):
        """Dumper without options return unmodified environment variables."""
        monkeypatch.setattr(dumper, 'environ', simple_environ())
        dump_result = dumper.dump()

        assert list(dump_result.keys()) == ['a', 'key']
        assert dump_result['key'] == 'value'
        assert dump_result['a'] == 'b'

    def test_with_prefix(self, monkeypatch):
        """Dumper with prefix option should return requested variables."""
        prefix = 'P_'
        monkeypatch.setattr(
            dumper,
            'environ',
            simple_environ(prefix=prefix),
        )
        dump_result = dumper.dump(prefixes=[prefix])

        assert len(dump_result.keys()) == 1
        assert dump_result['key'] == 'value'

    def test_with_template(self, monkeypatch, env_file):
        """Dumper with template option return variables for given template."""
        monkeypatch.setattr(dumper, 'environ', simple_environ())
        dump_result = dumper.dump(template=env_file)

        assert list(dump_result.keys()) == ['NORMAL_KEY', 'a', 'key']
        assert dump_result['key'] == 'value'
        assert dump_result['a'] == 'b'
        assert dump_result['NORMAL_KEY'] == 'SOMEVALUE'

    def test_with_two_options(self, monkeypatch, env_file):
        """Check both prefix and template options works together."""
        prefix = 'P_'
        monkeypatch.setattr(
            dumper,
            'environ',
            simple_environ(prefix=prefix),
        )
        dump_result = dumper.dump(template=env_file, prefixes=[prefix])

        assert list(dump_result.keys()) == ['NORMAL_KEY', 'key']
        assert dump_result['key'] == 'value'
        assert dump_result['NORMAL_KEY'] == 'SOMEVALUE'

    def test_with_multiple_prefixes(self, monkeypatch):
        """With multiple prefixes further prefixed variable replace previous."""
        first_prefix = 'P1_'
        monkeypatch.setattr(
            dumper,
            'environ',
            simple_environ(prefix=first_prefix),
        )
        second_prefix = 'P2_'
        monkeypatch.setattr(
            dumper,
            'environ',
            simple_environ(
                prefix=second_prefix,
                env_value='another_value',
            ),
        )
        dump_result = dumper.dump(prefixes=[first_prefix, second_prefix])

        assert len(dump_result.keys()) == 1
        assert dump_result['key'] == 'another_value'

    def test_dump_with_fill_argument(self, env_file_with_template):
        """Check fill arguments works."""
        dump_result = dumper.dump(template=env_file_with_template, fill=True)
        assert dump_result['VAR_TEMPLATE'] == 'value1 value2'


@pytest.mark.usefixtures('monkeypatch', 'env_file')
class TestDumpRegression:
    """Test dump function (regression)."""

    def test_same_environ(self, monkeypatch, env_file):
        """Dumper should return unmodified environment variables by default."""
        monkeypatch.setattr(
            dumper,
            'environ',
            same_environ(),
        )
        dump_result = dumper.dump(template=env_file)

        # Should contain the value from env, not from template:
        assert dump_result['NORMAL_KEY'] == 'test'

    def test_multiple_vars_with_prefix(self, monkeypatch, env_file, snapshot):
        """Dumper with prefix option should return all prefixed variables."""
        monkeypatch.setattr(
            dumper,
            'environ',
            multiple_variables_with_prefix(),
        )
        dump_result = dumper.dump(template=env_file, prefixes=['SECRET_'])

        # Only prefix should be changed, other parts should not:
        assert dump_result == snapshot
