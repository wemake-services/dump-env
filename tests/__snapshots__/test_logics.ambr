# serializer version: 1
# name: TestDumpRegression.test_multiple_vars_with_prefix
  OrderedDict({
    'DJANGO_SECRET_KEY': 'test',
    'NORMAL_KEY': 'SOMEVALUE',
    'SECRET_VALUE': 'value',
  })
# ---
